src/crypto/CHIPCryptoPAL.cpp

```diff
@@ -648,5 +649,185 @@ CHIP_ERROR GenerateCompressedFabricId(const Crypto::P256PublicKey & root_public_
+    if (!firmwareInfo.empty())
+    {
+        ReturnErrorOnFailure(tlvWriter.Put(TLV::ContextTag(4), firmwareInfo));
+    }
+    if (!vendorReserved5.empty())
+    {
+        ReturnErrorOnFailure(tlvWriter.Put(TLV::ContextTag(5), vendorReserved5));
+    }
+    if (!vendorReserved6.empty())
+    {
+        ReturnErrorOnFailure(tlvWriter.Put(TLV::ContextTag(6), vendorReserved6));
+    }
+    if (!vendorReserved7.empty())
+    {
+        ReturnErrorOnFailure(tlvWriter.Put(TLV::ContextTag(7), vendorReserved7));
+    }
+    if (!vendorReserved8.empty())
+    {
+        ReturnErrorOnFailure(tlvWriter.Put(TLV::ContextTag(8), vendorReserved8));
+    }
+    ReturnErrorOnFailure(tlvWriter.EndContainer(outerContainerType));
+    ReturnErrorOnFailure(tlvWriter.Finalize());
+    attestationElements = attestationElements.SubSpan(0, tlvWriter.GetLengthWritten());
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR DeconstructAttestationElements(const ByteSpan & attestationElements, ByteSpan & certificationDeclaration,
+                                          ByteSpan & attestationNonce, uint32_t & timestamp, ByteSpan & firmwareInfo,
+                                          ByteSpan & vendorReserved5, ByteSpan & vendorReserved6, ByteSpan & vendorReserved7,
+                                          ByteSpan & vendorReserved8)
+{
+    ByteSpan * element_array[] = { &certificationDeclaration, &attestationNonce, nullptr,          &firmwareInfo,
+                                   &vendorReserved5,          &vendorReserved6,  &vendorReserved7, &vendorReserved8 };
+
+    uint32_t validArgumentCount = 0;
+    uint32_t currentDecodeTagId = 0;
+    CHIP_ERROR TLVError         = CHIP_NO_ERROR;
+    CHIP_ERROR TLVUnpackError   = CHIP_NO_ERROR;
+    bool argExists[9];
+    TLV::TLVReader tlvReader;
+    TLV::TLVType containerType = TLV::kTLVType_Structure;
+
+    tlvReader.Init(attestationElements.data(), static_cast<uint32_t>(attestationElements.size()));
+    ReturnErrorOnFailure(tlvReader.Next(containerType, TLV::AnonymousTag));
+    ReturnErrorOnFailure(tlvReader.EnterContainer(containerType));
+
+    memset(argExists, 0, sizeof argExists);
+
+    for (size_t i = 0; i < sizeof(element_array) / sizeof(*element_array); ++i)
+    {
+        if (element_array[i] != nullptr)
+        {
+            *element_array[i] = ByteSpan();
+        }
+    }
+
+    while ((TLVError = tlvReader.Next()) == CHIP_NO_ERROR)
+    {
+        // Since call to aDataTlv.Next() is CHIP_NO_ERROR, the read head always points to an element.
+        // Skip this element if it is not a ContextTag, not consider it as an error if other values are valid.
+        if (!TLV::IsContextTag(tlvReader.GetTag()))
+        {
+            continue;
+        }
+        currentDecodeTagId = TLV::TagNumFromTag(tlvReader.GetTag());
+        if (currentDecodeTagId < 9 && currentDecodeTagId > 0)
+        {
+            if (argExists[currentDecodeTagId])
```

Comment:
It's odd that the first item is skipped in here.

Overall, it's not obvious at all how schema is validated for correctness in this method